python/python/Ice/Py3/IceFuture.py

```diff
@@ -19,25 +20,31 @@ def __await__(self):
             yield self
         return self.result()
 
+
 def wrap_future(future, *, loop=None):
     '''Wrap Ice.Future object into an asyncio.Future.'''
     if isinstance(future, asyncio.Future):
         return future
 
     assert isinstance(future, FutureBase), 'Ice.Future is expected, got {!r}'.format(future)
 
-    if loop is None:
-        loop = asyncio.get_event_loop()
-
-    af = asyncio.Future()
-
-    def callback():
-        if future.cancelled():
-            af.cancel()
-        elif future.exception():
-            af.set_exception(future.exception())
-        else:
-            af.set_result(future.result())
-
-    future.add_done_callback(lambda f: loop.call_soon_threadsafe(callback))
-    return af
+    def forwardCompletion(sourceFuture, targetFuture):
+        if not targetFuture.done():
+            if sourceFuture.cancelled():
+                targetFuture.cancel()
+            elif sourceFuture.exception():
+                targetFuture.set_exception(sourceFuture.exception())
+            else:
+                targetFuture.set_result(sourceFuture.result())
+
+    asyncioFuture = asyncio.Future()
+
+    if future.done():
+        forwardCompletion(future, asyncioFuture)
```

Comment:
This is thread-safe because it's ok to complete an asyncio future that nobody else has from a different thread?
Would be nice to add a comment.

Reply:
Hmm the future is created and completed on the event loop thread. How could this not be thread-safe?

Reply:
I don't understand your comment. It's created and completed on the thread that calls `Ice.wrap_future`.

Reply:
Ok I understand now. You're questioning the thread safety of the future completion for the case where `Ice.wrap_future` is called with another loop than the thread's running loop (`loop != None`). 

If the future implementation did rely on shared state which is only safe to access from the future's loop, I guess it would indeed be problematic but from looking at the implementation I don't think that's the case... as long as no "done callback"s are registered with the future, completing the future just mutate some internal state.  I'll add a comment.